A96-retry-otel-stats.md

```diff
@@ -0,0 +1,95 @@
+## A96: OTel Metrics for Retries
+
+*   Author: Yash Tibrewal (@yashykt)
+*   Approver: Mark Roth (@markdroth)
+*   Status: In Review
+*   Implemented in:
+*   Last updated: 2025-04-02
+*   Discussion at: https://groups.google.com/g/grpc-io/c/bFUHkcBA9cw
+
+## Abstract
+
+Propose OpenTelemetry metrics for gRPC retries.
+
+## Background
+
+Since OpenCensus has been sunsetted in favor of OpenTelemetry, gRPC has been
+developing its own OpenTelemetry plugin that is meant to replace the OpenCensus
+plugin ([A66]). This document proposes the OpenTelemetry version of the retry
+metrics originally proposed in [A45].
+
+### Related Proposals:
+
+*   [A45]: Exposing OpenCensus Metrics and Tracing for gRPC retry
+*   [A66]: OpenTelemetry Metrics
+*   [A79]: Non-per-call Metrics Architecture
+
+[A45]: A45-retry-stats.md
+[A66]: A66-otel-stats.md
+[A79]: A79-non-per-call-metrics-architecture.md
+
+## Proposal
+
+Metric Name                  | Type      | Unit    | Labels                                                                     | Description
+---------------------------- | --------- | ------- | -------------------------------------------------------------------------- | -----------
+grpc.client.call.retries     | Histogram | {retry} | grpc.method (required), grpc.target (required), grpc.retry_type (required) | Number of retry attempts made during the client call. The original attempt is not counted as a retry attempt. Recommended histogram bucket boundaries are [0,1,2,3,4,5,10,100,1000].
```

Comment:
`grpc.client.call.attempts` and `grpc.attempt_type`?

Hedging is not a kind of retry.

But..do you _need_ it to be 0 for "normal" cases without any retry/hedging?  Because calling it `attempts` would mean it should be 1 in those cases.

Reply:
> Hedging is not a kind of retry.

https://github.com/grpc/proposal/blob/master/A45-retry-stats.md counted hedging stats under retries. I'm fine calling it as an attempt instead.

well, thinking about it, we could just make `grpc.attempt_type` an optional attribute on the existing metric `grpc.client.attempt.started`.

Reply:
ok, my bad, I clearly forgot that `grpc.client.attempt.started` is a counter and not a histogram. This change will not work. To avoid any confusion on hedges being counted as retries, I am going to just use three different metrics for hedges, retries and transparent retries.

Reply:
> https://github.com/grpc/proposal/blob/master/A45-retry-stats.md counted hedging stats under retries. I'm fine calling it as an attempt instead.

For reference, we decided to not call it an attempt since we do not want to count original attempts. Those can be figured out by `grpc.client.attempt.started` - `grpc.client.call.retries` - `grpc.client.call.transparent_retries` - `grpc.client.call.hedges`